### DISCUSSION TITLE
How to setup an External Code Location

### DISCUSSION THREAD

**Discussion by broomva** [dagster-io/dagster#20255] [Q&A]
I want to set up a folder (s3 or similar bucket storage) with various .py files to be used as dagster pipelines. Is that something that can be readily used?

**Comment by garethbrickman** [comment]
There is an experimental feature called [External Assets](https://docs.dagster.io/concepts/assets/external-assets#external-assets-experimental) that I think fits your use case

**Comment by broomva** [comment]
Huh, not quite. What I meant is to have a code location somewhere else besides the gprc defined by the daemon. The question arises since I'm having trouble running dagster on a cloud service (railway app), either as a python or docker deployment. Check out the error for the python attempt:

`/usr/local/lib/python3.11/site-packages/dagster/_core/workspace/context.py:622: UserWarning: Error loading repository location web_rag_agent.py:Exception: gRPC server exited with return code -9 while starting up with the command: "/usr/local/bin/python -m dagster api grpc --lazy-load-user-code --socket /tmp/tmphmb1ukpe --heartbeat --heartbeat-timeout 45 --fixed-server-id 63dd9761-92fe-4102-8671-a6b1f8fe1475 --log-level info --inject-env-vars-from-instance --instance-ref {"__class__": "InstanceRef", "compute_logs_data": {"__class__": "ConfigurableClassData", "class_name": "LocalComputeLogManager", "config_yaml": "base_dir: /root/storage\n", "module_name": "dagster.core.storage.local_compute_log_manager"}, "custom_instance_class_data": null, "event_storage_data": {"__class__": "ConfigurableClassData", "class_name": "SqliteEventLogStorage", "config_yaml": "base_dir: /root/history/runs/\n", "module_name": "dagster.core.storage.event_log"}, "local_artifact_storage_data": {"__class__": "ConfigurableClassData", "class_name": "LocalArtifactStorage", "config_yaml": "base_dir: /root/\n", "module_name": "dagster.core.storage.root"}, "run_coordinator_data": {"__class__": "ConfigurableClassData", "class_name": "DefaultRunCoordinator", "config_yaml": "{}\n", "module_name": "dagster.core.run_coordinator"}, "run_launcher_data": {"__class__": "ConfigurableClassData", "class_name": "DefaultRunLauncher", "config_yaml": "{}\n", "module_name": "dagster"}, "run_storage_data": {"__class__": "ConfigurableClassData", "class_name": "SqliteRunStorage", "config_yaml": "base_dir: /root/history/\n", "module_name": "dagster.core.storage.runs"}, "schedule_storage_data": {"__class__": "ConfigurableClassData", "class_name": "SqliteScheduleStorage", "config_yaml": "base_dir: /root/schedules\n", "module_name": "dagster.core.storage.schedules"}, "scheduler_data": {"__class__": "ConfigurableClassData", "class_name": "DagsterDaemonScheduler", "config_yaml": "{}\n", "module_name": "dagster.core.scheduler"}, "secrets_loader_data": null, "settings": {}, "storage_data": {"__class__": "ConfigurableClassData", "class_name": "DagsterSqliteStorage", "config_yaml": "base_dir: /root/\n", "module_name": "dagster.core.storage.sqlite_storage"}} --location-name web_rag_agent.py -f /opt/dagster/app/dags/web_rag_agent.py"`



Or with a docker image:
```
/usr/local/lib/python3.11/site-packages/dagster/_core/workspace/context.py:622: UserWarning: Error loading repository location vortex_demo:dagster._core.errors.DagsterUserCodeUnreachableError: Could not reach user code server. gRPC Error code: UNAVAILABLE

Stack Trace:

File "/usr/local/lib/python3.11/site-packages/dagster/_core/workspace/context.py", line 617, in _load_location

else origin.create_location(self.instance)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "/usr/local/lib/python3.11/site-packages/dagster/_core/host_representation/origin.py", line 373, in create_location

return GrpcServerCodeLocation(self, instance=instance)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "/usr/local/lib/python3.11/site-packages/dagster/_core/host_representation/code_location.py", line 632, in __init__

list_repositories_response = sync_list_repositories_grpc(self.client)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "/usr/local/lib/python3.11/site-packages/dagster/_api/list_repositories.py", line 20, in sync_list_repositories_grpc

api_client.list_repositories(),

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "/usr/local/lib/python3.11/site-packages/dagster/_grpc/client.py", line 250, in list_repositories

res = self._query("ListRepositories", api_pb2.ListRepositoriesRequest)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

File "/usr/local/lib/python3.11/site-packages/dagster/_grpc/client.py", line 173, in _query

self._raise_grpc_exception(

File "/usr/local/lib/python3.11/site-packages/dagster/_grpc/client.py", line 156, in _raise_grpc_exception

raise DagsterUserCodeUnreachableError(

The above exception was caused by the following exception:

grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:

status = StatusCode.UNAVAILABLE

details = "failed to connect to all addresses; last error: UNKNOWN: ipv4:127.0.0.1:4266: Failed to connect to remote host: Connection refused"

debug_error_string = "UNKNOWN:Error received from peer  {grpc_message:"failed to connect to all addresses; last error: UNKNOWN: ipv4:127.0.0.1:4266: Failed to connect to remote host: Connection refused", grpc_status:14, created_time:"2024-03-05T06:16:30.664136417+00:00"}"
```

You may see the [dockerfile here](https://github.com/Broomva/vortex/blob/main/examples/vortex_demo/Dockerfile)

**Reply by garethbrickman** [reply]
Ah my apologies for misunderstanding! By default, Dagster tools automatically create a process on your local machine for each of your code locations. However, it's also possible to run your own gRPC server that's responsible for serving information about your code locations. This can be useful in more complex system architectures that deploy user code separately from the Dagster webserver. 

More info about that [here](https://docs.dagster.io/concepts/code-locations/workspace-files#running-your-own-grpc-server). So you would have to configure wherever your own gRPC server is hosted to fetch the needed files from external storage.

**Comment by ismailsimsek** [comment]
+1, also interested in defining/using S3 code location as a code location. is this possible? 

the idea is deploying the dagster project to S3 with CI/CD then Dagster is loading the projects from from S3

**Reply by ion-elgreco** [reply]
If you are using dagster dev -f <>, you could sync/mount the s3 bucket files to local fs and then point to the relative filepath of the mount.

**Reply by anentropic** [reply]
if the files in the bucket change do I have to restart the server?

I'd ideally like to see something like AWS MWAA offers for Airflow, where DAGs are hot reloaded from S3 bucket:
https://docs.aws.amazon.com/mwaa/latest/userguide/configuring-dag-folder.html
